Document sum and usum (#352)
Certora/Documentation: docs/cvl/expr.md

{ref}`grounding` for more detail.
```

* The prover is capable of calculating the `sum` of ghost mappings over
specified indices. For example, if we have a ghost mapping `ghost
mapping(address => mapping(bytes32 => mathint)) myGhost`, and want to sum all
the values of the ghost over all addresses for a given `bytes32` value `b`, one

> could write `sum address a. myGhost[a][b]`. This will return a `mathint`-typed

Reviewer: ```suggestion
can write `sum address a. myGhost[a][b]`. This returns a `mathint`-typed
```

number that sums all known values of `myGhost` over the first index.
The full syntax for this is `sum type1 t1, type2 t2, ... typeN tN.
ghostName[...]`

```{note}
The prover support only summation of ghosts. If one wants to sum e.g. some
storage mapping, one could implement a ghost that mirrors the storage and sum
over it.

* An extension of the summing logic is the unsigned sum which uses the `usum`
keyword. It follows the same rules as the regular sum, but adds some extra
logic to ensure the value of the sum is always larger than its parts.

```{note}
A `usum` can only be used on ghost mappings with unsigned (or `mathint`) value
types. In the case of `mathint` an assertion is added on each write to the ghost

> that the written value is indeed unsigned.

Reviewer: Can we explain the difference between `usum` and `sum` better here? In particular, what's the difference of `sum` over an unsigned ghost mapping vs `usum` over the same mapping? @jhoenicke can you take a look and make a suggestion here?

Reviewer 2: Maybe we can just add this

```
The keyword `usum` indicates that all entries are non-negative (unsigned). 
It can only be used on ghost mappings with unsigned (or `mathint`) value
types. In the case of `mathint` an assertion is added on each write to the ghost
that reports an error if the written value is negative.
When using this keyword, the solver will introduce the additional assumption
that the `usum` is larger than any value in the ghost mapping and that it is 
even larger than the sum of any finite subset of values.
This additional assumption is valid, because the other values in the ghost
mapping can make the total sum only larger. 
```

It's a bit repeating what was written above it, but maybe it's okay to repeat that here.

{ref}`grounding` for more detail.
```

* The prover is capable of calculating the `sum` of ghost mappings over
specified indices. For example, if we have a ghost mapping `ghost
mapping(address => mapping(bytes32 => mathint)) myGhost`, and want to sum all
the values of the ghost over all addresses for a given `bytes32` value `b`, one
can write `sum address a. myGhost[a][b]`. This returns a `mathint`-typed
number that sums all known values of `myGhost` over the first index.
The full syntax for this is `sum type1 t1, type2 t2, ... typeN tN.

> ghostName[...]`. See [here](https://github.com/Certora/Examples/blob/master/CVLByExample/Summarization/GhostSummary/GhostSums/README.md)

Reviewer: @yoav-el-certora fyi - we are hard-coding the link to the example on `master` in the docs here (even though it hasn't been merged). This breaks if the files are moved. Please ensure to keep the example in this location.
